SONiC Audit Enhancement

maipbui opened the pull request:
This design aims to enhance the auditing capabilities within SONiC operating system using audit daemon (auditd). Auditing is the process of recording and analyzing the events that occur on the device. Auditing can help to detect unauthorized access, configuration changes, malicious activity, or system errors. Auditing can also provide evidence for forensic investigations, compliance audits, or incident response.

qiluo-msft commented on doc/audit/SONiC_Audit_Enhancement.md:
````diff
@@ -0,0 +1,149 @@
+
+```
+admin@sonic# config audit
+Usage: config [OPTIONS] COMMAND [ARGS]...
````
>config [](http://example.com/codeflow?start=7&length=6)

You should also define `show` command

maipbui replied:
Updated

qiluo-msft commented on doc/audit/SONiC_Audit_Enhancement.md:
```diff
@@ -0,0 +1,149 @@
+| Log changes              | `-w /var/log -p wa -k log_changes` |
+| Docker related           | `-w /usr/bin/dockerd -p wa -k docker_daemon`<br>`-w /etc/docker/daemon.json -p wa -k docker_config`<br>`-w /lib/systemd/system/docker.service -p wa -k docker_service`<br>`-w /lib/systemd/system/docker.socket -p wa -k docker_socket`<br>`-a always,exit -F arch=b64 -S execve -F path=/usr/bin/docker -k docker_commands`<br>`-w /var/lib/docker/ -p wa -k docker_storage`<br>` ...
+
+### 3.3 CLI design
```
>3.3 [](http://example.com/codeflow?start=4&length=3)

Do you have ConfigDB new schema design? If yes, you also need to define for the warm-reboot requirement. Please check [https://github.com/sonic-net/SONiC/blob/master/doc/hld_template.md](https://github.com/sonic-net/SONiC/blob/master/doc/hld_template.md) if any sections are missing.

maipbui replied:
Updated

maipbui commented on doc/audit/SONiC_Audit_Enhancement.md:
```diff
@@ -0,0 +1,149 @@
+
+- Introduce a new file, `custom-audit.rules`, into the `/etc/audit/rules.d/` directory.
+- Additionally, the `/etc/audit/plugins.d/syslog.conf` is modified (by setting `active = yes`) to enable sending auditd logs to a syslog server.
+- CLI commands enhancement:
```
Updated

maipbui commented on doc/audit/SONiC_Audit_Enhancement.md:
```diff
@@ -0,0 +1,149 @@
+###### Table 2: Rules Review
+| Rule name | Details |
+|--------------------------|--------------------------|
+| Process audit            | `-a never,exit -F path=/usr/bin/runc -F perm=x`<br>`-a never,exit -F path=/usr/bin/python* -F perm=x`<br>`-a never,exit -F path=/usr/bin/docker -F perm=x` |
```
We don't need to write the name in `config` CLI, `config audit` - enables/disables all audit rules including existing `audisp-tacplus.rules` and new `custom-audit.rules`

qiluo-msft commented on doc/audit/security_auditing_HLD.md:
```diff
@@ -0,0 +1,366 @@
+  - [Table of Content](#table-of-content)
+  - [List of Tables](#list-of-tables)
+  - [Revision](#revision)
+          - [Table 1: Revision](#table-1-revision)
```
You can remove table from TOC

maipbui replied:
Updated

qiluo-msft commented on doc/audit/security_auditing_HLD.md:
````diff
@@ -0,0 +1,366 @@
+```
+{
+    "AUDIT": {
+      "config": {
````
Could you design for group-level control, such as crtical file change, DNS change, etc? Different enterprises may require different set of audit rules.

maipbui replied:
Updated

qiluo-msft commented on doc/audit/security_auditing_HLD.md:
```diff
@@ -0,0 +1,366 @@
+
+| Key       | Description                 |
+| --------- | --------------------------- |
+| enabled   | The flag indicating whether the AUDIT enabled, 1 enabled, others not enabled  |
```
Is this table in StateDB? I guess you miss some description paragraph.

maipbui replied:
Updated, table will be in ConfigDB

zhangyanzhao commented:
community review recording https://zoom.us/rec/share/I2O2drYuFPNLFyWGPanI_rQ7qPIOkyfa9yNyJSON7BxDbKWcshSltjHwfUueU_tN.EySn3jaIk96H4zXA

qiluo-msft commented on doc/audit/security_auditing_HLD.md:
````diff
@@ -0,0 +1,366 @@
+#### 3.3.4 Logrotate
+The following settings in the `/etc/logrotate.d/audit` file set up log rotation for audit logs:
+```
+{
````
Do you need to specify the log filename?

maipbui replied:
Updated

maipbui commented:
According to Sonic team and community's feedback, addressed and added some additional topics to the HLD
- Add Config DB improvement
- Add YANG section
- Add Audit rule ordering
- Add test case for ordering of rules
- Add test case for performance test
- Add CPU/MEM utilization section
- Support fine-grained control to add or remove an individual rule
- Add Q&A, future improvements section

xincunli-sonic commented on doc/audit/security_auditing_HLD.md:
```diff
@@ -0,0 +1,513 @@
+| 0.1 | 08/20/2024  | Mai Bui            | Initial version                   |
+
+## Scope
+This section describes the audit enhancement high-level design in SONiC.
```
Please elaborating more on the scope of the document. What specific security aspects are being audited?

maipbui replied:
Updated

xincunli-sonic commented on doc/audit/security_auditing_HLD.md:
```diff
@@ -0,0 +1,513 @@
+This document outlines the addition of several auditd rules to enhance the auditing capabilities on a SONiC device. These extra rules aim to provide thorough monitoring of important system activities, making the device more secure and better prepared for compliance and incident response.
+
+## 2. Current Architecture Design
+In SONiC, audit settings are centrally managed through a configuration file at `/etc/audit/auditd.conf`. Upon system startup, SONiC automatically compiles audit rules from the `/etc/audit/rules.d/` directory into a single file, `/etc/audit/audit.rules`, which auditd then loads. Here is a brief look at the relevant files:
```
Would be better if have a diagram for current architecture?

maipbui replied:
Good idea, but actually, I added proposed architecture diagram, I think this way is better

xincunli-sonic commented on doc/audit/security_auditing_HLD.md:
```diff
@@ -0,0 +1,513 @@
+### 3.2 Audit Rules Review
+###### Table 2: Audit Rules Review
+These rules will be included in the image and enabled by default.
+| Rule name | Details |
```
Could we add a column by providing examples of the audit rules?

maipbui replied:
I added some audit logs examples, it looks similar, so I did not add entire column

xincunli-sonic commented on doc/audit/security_auditing_HLD.md:
```diff
@@ -0,0 +1,513 @@
+                    description "List of audit rules";
+
+                    leaf name {
+                        type string;
```
In schema you defined length, how about add length here to restrict?

maipbui replied:
Updated

xincunli-sonic commented on doc/audit/security_auditing_HLD.md:
```diff
@@ -0,0 +1,513 @@
+
+                    list rule {
+                        key "rule";
+                        type string;
```
Suggest adding length.

maipbui replied:
Updated

xincunli-sonic commented on doc/audit/security_auditing_HLD.md:
```diff
@@ -0,0 +1,513 @@
+
+        container AUDIT {
+
+            description "AUDIT part of config_db.json";
```
remove .json?

maipbui replied:
Updated

maipbui commented on doc/audit/security_auditing_HLD.md:
```diff
@@ -0,0 +1,513 @@
+enable                       = boolean                      ; Indicates whether security auditing is enabled (true/false)
+RULESET                      = list                         ; List of audit rule sets
+name                         = 1*255VCHAR                   ; Name of the audit rule set
+rule                         = 1*255VCHAR                   ; Audit rule definition in auditd format
```
This won't be support for now, I added to future improvements

xincunli-sonic reviewed (approved):
LGTM

qiluo-msft commented on doc/audit/security_auditing_HLD.md:
```diff
@@ -37,7 +37,7 @@
 * [Table 1: Revision](#table-1-revision)
 * [Table 2: Audit Rules Review](#table-2-audit-rules-review)
-* [Table 3: Unit Test Cases](#table-3-unit-test-cases)
+* [Table 3: Unt Test Cases](#table-3-unit-test-cases)
```
typo: Unt

maipbui replied:
Fixed

maipbui commented on doc/audit/security_auditing_HLD.md:
```diff
@@ -87,89 +91,111 @@ In SONiC, audit settings are centrally managed through a configuration file at `
-                    ]
-                }
-            ]
+            "critical_files": "enabled",
```
Fixed, thanks!

maipbui commented:
@liuh-80 @Yarden-Z @xincunli-sonic @venkatmahalingam @yxieca @prsunny @lguohan could you please help re-review the HLD? thanks

zhangyanzhao commented:
@reviewers, would you please help to review this security feature? Thanks.

zhangyanzhao commented:
PRs are not merged yet, move to backlog
